### ISSUE TITLE
Getting ready for Phoenix 1.3

### ISSUE THREAD

**Issue by Qqwy** [smpallen99/coherence#176]
In Phoenix 1.3, the `web` folder is moved under the `lib` folder. This meant that after running the generator, I had to rename some things:

1. Merge the `./web` folder with the `./lib/web` folder.
2. In `./lib/web/coherence_web.ex`, rename:
`MyApp.Router.Helpers` -> `MyApp.Web.Router.Helpers`
`MyApp.ErrorHelpers` -> `MyApp.Web.ErrorHelpers`
`MyApp.Gettext` -> `MyApp.Web.Gettext`

3. Change `use Phoenix.View, root: "web/templates/coherence"` -> `use Phoenix.View, root: "lib/myapp/web/templates/coherence"`

4. The scopes that are mentioned you should add to your `router.ex` schould be `MyApp.Web` instead of `MyApp`.

5. In the generated User model, `./lib/web/models/user.ex`, instead of `use MyApp.Web, :model`, use `use Ecto.Schema`.

6. In `./lib/web/views/coherence/coherence_view_helpers.ex`, change `  @helpers Module.concat(Application.get_env(:coherence, :module), Router.Helpers)` to `  @helpers Module.concat(Application.get_env(:coherence, :module), Web.Router.Helpers)`

I believe that is all :slightly_smiling_face: !

**Comment by smpallen99** [comment]
@Qqwy have you looked at the phx-1.3 branch? I believe i've done most of the work to work in user's project. I have not restructured the internals of coherence yet, but that should not affect coherence working with phx projects.

I have not merged into master because the branch is not backwards compatible with pre 1.3 projects yet.

**Comment by smpallen99** [comment]
I'm working on a phx 1.3 umbrella project now. I've started restructuring the internals based on the new phx architecture. This is not a simple task. There are two different project structures for `phx.new`, `phx.new.web`, and `phx.new.ecto`, as well as backwards compatibility for the legacy project structure.  Furthermore, the I don't believe the phx generators are complete yet.

**Comment by tmbb** [comment]
This question comes from someone who first heard about elixir about 2 to 3 weeks ago, so I apologize in advance if the answer turns out to be very obvious. I still haven't got the hang of this whole OTP app stuff, and I'm using Phoenix as if it were Django with channels. I've looked at the code (both the master and the phx-1.3 branch) and apart from the template structure not much seems to have changed. This means that the new branch is compatible with apps built on Phoenix 1.2, right?

**Comment by agustif** [comment]
@tmbb I think the answer is yes, if you look up in some other issues I remember @smpallen99 saying he was waiting to push 1.3 patches to master because of some incompatibility issues with 1.2 but also some stuff on the generators on RC not being complete yet.

I think it's a clear choice, if you don't have a hurry just wait to 1.3 to be more mature, it won't take more than a few weeks for major libraries to catch up I think.

If you're in a hurry, you're perfectly good building a deployable app with Phoenix 1.2/Elixir 1.3, and just wait it out to clear update paths.... Maybe you can build 1.2 even having in mind the new 1.3 structures and stuff, but it's not that relevant I think.

**Comment by tmbb** [comment]
Well, I won't update my existing app to Phoenix 1.3 in the near future. I was asking because if my new PR ends up merged into the Phx 1.3 branch, I won't be able to make use of the new infernationalization features. I'll probably just declare my personal repo as a dependency while I wait for new developments on the Phx 1.3 branch.

**Comment by smpallen99** [comment]
The phx-1.3 branch will not support projects in the old file structure yet. I hard coded some of the file locations to get something to support the new structure quickly. I’m part way through the changes to detect the project structure. However, this is not real easy. There are about  4 different combinations of project structures I need to handle.

**Comment by smpallen99** [comment]
What do you guys think about adding new `coh` mix tasks for support of projects build with `mix phx.new`. This way, we can reduce the complexity of the install task and and keep the old `mix coherence.install` for the older projects?

**Comment by smpallen99** [comment]
I've decided to use the `coh.install` approach. Finished most of the work. Tests are passing. Just doing some manual integration testing on real projects.

**Comment by smpallen99** [comment]
The latest push to this branch now support both legacy and phx structured project. Still todo:

- [ ]  Test both of the `clean` mix tasks
- [ ]  Test the reinstall options on both install tasks
- [ ]  Move the `coherence schema out of web.
- [ ]  Namespace the web components with `Web`
- [ ]  Add support for umbrella apps created with `phx.new.web` and `phx.new.ecto`

**Comment by egeersoz** [comment]
Is there a release date target for this branch?

**Comment by hakunin** [comment]
@egeersoz Just installed coherence from the 1.3 branch and it works nicely

**Comment by Qqwy** [comment]
I just tried installing Coherence from the `phx-1.3` branch, and running `mix coherence.install --full-confirmable` results in a root-level `web` folder to be made (rather than the coherence files being added to `lib/web`). Not sure about the other things that might be missing, but it is clearly not done yet :-).

**Comment by hakunin** [comment]
I can confirm that I had to move the web folder as well.

**Comment by tmbb** [comment]
@Qqwy and @hakunin you're getting the expected results...

From the [README](https://github.com/smpallen99/coherence/tree/phx-1.3#coherence) of the phx-1.3 branch:

> For backwards compatiblity, the `mix coherence.install` and `mix coherence.clean` remain the same. For projects that have converted the new 1.3 project structure, please use `mix coh.install` and `mix coh.clean`.

Have you tried `mix coh.install`?

**Comment by Qqwy** [comment]
Turns out I did two things wrong:

1. I used `{:coherence, github: "smpallen99/coherence", rel: "phx-1.3"}` rather than `{:coherence, github: "smpallen99/coherence", branch: "phx-1.3"}` as dependency, which happily grabbed some older version that did not have the `coh.install` task for me.
2. I did not know that there was a different installation task.

It would be swell if `coherence.install` would warn you (with a continue? y/n option) if it is run on a Phoenix 1.3 project.


Thank you, @tmbb !
